python/sglang/srt/models/deepseek_v2.py

```diff
@@ -1281,18 +1306,30 @@ def forward_ffn_with_full_input(
             )
 
         # Fully Connected
-        hidden_states = self.mlp(hidden_states)
+        hidden_states = self.mlp(
+            hidden_states, split_indices_cpu=forward_batch.split_indices_cpu
+        )
 
         # TODO(ch-wan): ues reduce-scatter in MLP to avoid this scatter
         # Scatter
         if self.dp_size != 1:
             # important: forward batch.gathered_buffer is used both after scatter and after gather.
             # be careful about this!
-            hidden_states, global_hidden_states = (
-                forward_batch.gathered_buffer[: forward_batch.input_ids.shape[0]],
-                hidden_states,
-            )
-            dp_scatter(hidden_states, global_hidden_states, forward_batch)
+            if not self.info.is_sparse or self.dp_size != self.tp_size:
+                hidden_states, global_hidden_states = (
+                    forward_batch.gathered_buffer[: forward_batch.input_ids.shape[0]],
+                    hidden_states,
+                )
+                dp_scatter(hidden_states, global_hidden_states, forward_batch)
+            else:
+                if hidden_states.shape[0] != residual.shape[0]:
+                    hidden_states, tmp_hidden_states = (
+                        forward_batch.gathered_buffer[
+                            : forward_batch.input_ids.shape[0]
+                        ],
+                        hidden_states,
+                    )
+                    dp_copy(hidden_states, tmp_hidden_states, forward_batch)
```

Comment:
If my understanding is correct, this line is doing the same thing with `dp_scatter`.